File: Problems/0385-mini-parser.py
# Mock NestedInteger for local testing

class NestedInteger(object):
    def __init__(self, value=None):
        if value is None:
            self.value = None
            self.list = []
        else:
            self.value = value
            self.list = None

    def isInteger(self):
        return self.value is not None

    def add(self, elem):
        if self.list is None:
            self.list = []
            self.value = None
        self.list.append(elem)

    def setInteger(self, value):
        self.value = value
        self.list = None

    def getInteger(self):
        return self.value

    def getList(self):
        return self.list


# LeetCode solution starts from here

class Solution(object):
    def deserialize(self, s):
        if s[0] != '[':
            return NestedInteger(int(s))

        stack = []
        num = 0
        sign = 1
        has_num = False

        for c in s:
            if c == '[':
                stack.append(NestedInteger())
            elif c == '-':
                sign = -1
            elif '0' <= c <= '9':
                num = num * 10 + ord(c) - 48
                has_num = True
            elif c == ',' or c == ']':
                if has_num:
                    stack[-1].add(NestedInteger(sign * num))
                    num = 0
                    sign = 1
                    has_num = False
                if c == ']' and len(stack) > 1:
                    cur = stack.pop()
                    stack[-1].add(cur)

        return stack[0]

File: Problems/0385-mini-parser.cpp
#include <bits/stdc++.h>
using namespace std;

class NestedInteger {
    bool isInt;
    int value;
    vector<NestedInteger> list;

public:
    NestedInteger() {
        isInt = false;
        value = 0;
    }

    NestedInteger(int value) {
        isInt = true;
        this->value = value;
    }

    bool isInteger() const {
        return isInt;
    }

    void add(const NestedInteger& elem) {
        if (isInt) {
            isInt = false;
            list.clear();
        }
        list.push_back(elem);
    }

    void setInteger(int value) {
        isInt = true;
        this->value = value;
        list.clear();
    }

    int getInteger() const {
        return value;
    }

    const vector<NestedInteger>& getList() const {
        return list;
    }
};

class Solution {
public:
    NestedInteger deserialize(string s) {
        if (s[0] != '[')
            return NestedInteger(stoi(s));

        vector<NestedInteger> stack;
        int num = 0, sign = 1;
        bool hasNum = false;

        for (char c : s) {
            if (c == '[') {
                stack.push_back(NestedInteger());
            }
            else if (c == '-') {
                sign = -1;
            }
            else if (isdigit(c)) {
                num = num * 10 + (c - '0');
                hasNum = true;
            }
            else if (c == ',' || c == ']') {
                if (hasNum) {
                    stack.back().add(NestedInteger(sign * num));
                    num = 0;
                    sign = 1;
                    hasNum = false;
                }

                if (c == ']' && stack.size() > 1) {
                    NestedInteger cur = stack.back();
                    stack.pop_back();
                    stack.back().add(cur);
                }
            }
        }

        return stack[0];
    }
};
